File: app/src/main/java/mega/privacy/android/app/lollipop/qrcode/QRCodeActivity.java
package mega.privacy.android.app.lollipop.qrcode;

import android.Manifest;
import android.app.Activity;
import android.content.Intent;
import android.content.pm.PackageManager;
import android.net.Uri;
import android.os.Build;
import android.os.Bundle;
import android.os.StatFs;
import android.support.design.widget.TabLayout;
import android.support.v4.app.ActivityCompat;
import android.support.v4.content.ContextCompat;
import android.support.v4.content.FileProvider;
import android.support.v4.view.ViewPager;
import android.support.v4.widget.DrawerLayout;
import android.support.v7.app.ActionBar;
import android.support.v7.widget.Toolbar;
import android.util.DisplayMetrics;
import android.view.Display;
import android.view.Menu;
import android.view.MenuInflater;
import android.view.MenuItem;
import android.view.View;

import java.io.File;
import java.io.FileInputStream;
import java.io.FileOutputStream;
import java.io.IOException;
import java.nio.channels.FileChannel;

import mega.privacy.android.app.MegaApplication;
import mega.privacy.android.app.R;
import mega.privacy.android.app.lollipop.FileStorageActivityLollipop;
import mega.privacy.android.app.lollipop.ManagerActivityLollipop;
import mega.privacy.android.app.lollipop.PinActivityLollipop;
import mega.privacy.android.app.modalbottomsheet.QRCodeSaveBottomSheetDialogFragment;
import mega.privacy.android.app.utils.Constants;
import mega.privacy.android.app.utils.Util;
import nz.mega.sdk.MegaApiAndroid;
import nz.mega.sdk.MegaApiJava;
import nz.mega.sdk.MegaContactRequest;
import nz.mega.sdk.MegaError;
import nz.mega.sdk.MegaRequest;
import nz.mega.sdk.MegaRequestListenerInterface;

import static mega.privacy.android.app.utils.CacheFolderManager.buildQrFile;

public class QRCodeActivity extends PinActivityLollipop implements MegaRequestListenerInterface{

    private static int REQUEST_DOWNLOAD_FOLDER = 1000;
    private final int MY_PERMISSIONS_REQUEST_CAMERA = 1010;

    private Toolbar tB;
    private ActionBar aB;

    private MenuItem shareMenuItem;
    private MenuItem saveMenuItem;
    private MenuItem settingsMenuItem;
    private MenuItem resetQRMenuItem;
    private MenuItem deleteQRMenuItem;

    private TabLayout tabLayoutQRCode;
    private ViewPager viewPagerQRCode;

    private ScanCodeFragment scanCodeFragment;
    private MyCodeFragment myCodeFragment;

    private QRCodePageAdapter qrCodePageAdapter;

    private DrawerLayout drawerLayout;

    private int qrCodeFragment;

    MegaApiAndroid megaApi;

    private boolean contacts = false;
    private boolean inviteContacts = false;

    DisplayMetrics outMetrics;

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        log("onCreate");

        if (savedInstanceState != null) {
            contacts = savedInstanceState.getBoolean("contacts", false);
            inviteContacts = savedInstanceState.getBoolean("inviteContacts", false);
        }
        else {
            contacts = getIntent().getBooleanExtra("contacts", false);
            inviteContacts = getIntent().getBooleanExtra("inviteContacts", false);
        }

        Display display = getWindowManager().getDefaultDisplay();
        outMetrics = new DisplayMetrics();
        display.getMetrics(outMetrics);

        setContentView(R.layout.activity_qr_code);

        tB = (Toolbar) findViewById(R.id.toolbar);
        if(tB==null){
            log("Tb is Null");
            return;
        }

        if (megaApi == null) {
            megaApi = ((MegaApplication) getApplication()).getMegaApi();
        }

        tB.setVisibility(View.VISIBLE);
        setSupportActionBar(tB);
        aB = getSupportActionBar();
        aB.setHomeButtonEnabled(true);
        aB.setDisplayHomeAsUpEnabled(true);
        tB.setTitle(getString(R.string.section_qr_code));

        drawerLayout = (DrawerLayout) findViewById(R.id.drawer_layout);

        qrCodePageAdapter =new QRCodePageAdapter(getSupportFragmentManager(),this);

        tabLayoutQRCode =  (TabLayout) findViewById(R.id.sliding_tabs_qr_code);
        viewPagerQRCode = (ViewPager) findViewById(R.id.qr_code_tabs_pager);

        if (ContextCompat.checkSelfPermission(this, Manifest.permission.CAMERA) != PackageManager.PERMISSION_GRANTED) {
            ActivityCompat.requestPermissions(this, new String[]{Manifest.permission.CAMERA}, MY_PERMISSIONS_REQUEST_CAMERA);
        }else {
            initActivity();
        }
    }

    @Override
    public void onRequestPermissionsResult(int requestCode, String permissions[], int[] grantResults) {
        switch (requestCode) {
            case MY_PERMISSIONS_REQUEST_CAMERA: {
                // If request is cancelled, the result arrays are empty.
                if (grantResults.length > 0 && grantResults[0] == PackageManager.PERMISSION_GRANTED) {
                    initActivity();
                }else{
                    this.finish();
                }
                return;
            }
        }
    }

    void initActivity(){
        viewPagerQRCode.addOnPageChangeListener(new ViewPager.OnPageChangeListener() {
            @Override
            public void onPageScrolled(int position, float positionOffset, int positionOffsetPixels) {

            }

            @Override
            public void onPageSelected(int position) {

                supportInvalidateOptionsMenu();

                if (position == 0) {
                    qrCodeFragment = 0;
                    myCodeFragment = (MyCodeFragment) qrCodePageAdapter.instantiateItem(viewPagerQRCode, 0);
                }
                else {
                    qrCodeFragment = 1;
                    scanCodeFragment = (ScanCodeFragment) qrCodePageAdapter.instantiateItem(viewPagerQRCode, 1);
                }
            }

            @Override
            public void onPageScrollStateChanged(int state) {

            }
        });

        viewPagerQRCode.setAdapter(qrCodePageAdapter);
        tabLayoutQRCode.setupWithViewPager(viewPagerQRCode);
        if (contacts || inviteContacts){
            viewPagerQRCode.setCurrentItem(1);
        }
        else {
            viewPagerQRCode.setCurrentItem(0);
        }
    }

    @Override
    protected void onSaveInstanceState(Bundle outState) {
        super.onSaveInstanceState(outState);
        outState.putBoolean("contacts", contacts);
        outState.putBoolean("inviteContacts", inviteContacts);
    }

    @Override
    public boolean onCreateOptionsMenu(Menu menu) {
        log("onCreateOptionsMenu");

        MenuInflater inflater = getMenuInflater();
        inflater.inflate(R.menu.activity_qr_code, menu);

        shareMenuItem = menu.findItem(R.id.qr_code_share);
        saveMenuItem = menu.findItem(R.id.qr_code_save);
        settingsMenuItem = menu.findItem(R.id.qr_code_settings);
        resetQRMenuItem = menu.findItem(R.id.qr_code_reset);
        deleteQRMenuItem = menu.findItem(R.id.qr_code_delete);

        switch (qrCodeFragment) {
            case 0: {
                shareMenuItem.setVisible(true);
                saveMenuItem.setVisible(true);
                settingsMenuItem.setVisible(true);
                resetQRMenuItem.setVisible(true);
                deleteQRMenuItem.setVisible(true);
                break;
            }
            case 1: {
                shareMenuItem.setVisible(false);
                saveMenuItem.setVisible(false);
                settingsMenuItem.setVisible(false);
                resetQRMenuItem.setVisible(false);
                deleteQRMenuItem.setVisible(false);
                break;
            }
        }

//        if (contacts) {
////            ScanCodeFragment.scannerView.setAutoFocus(true);
////            ScanCodeFragment.scannerView.startCamera();
//        }

        return super.onCreateOptionsMenu(menu);
    }

    @Override
    public boolean onOptionsItemSelected(MenuItem item) {
        log("onOptionsItemSelected");

        int id = item.getItemId();
        switch(id) {
            case android.R.id.home: {
                onBackPressed();
                break;
            }
            case R.id.qr_code_share: {
                shareQR ();
                break;
            }
            case R.id.qr_code_save: {
                QRCodeSaveBottomSheetDialogFragment qrCodeSaveBottomSheetDialogFragment = new QRCodeSaveBottomSheetDialogFragment();
                qrCodeSaveBottomSheetDialogFragment.show(getSupportFragmentManager(), qrCodeSaveBottomSheetDialogFragment.getTag());

                break;
            }
            case R.id.qr_code_settings: {
                Intent intent = new Intent(this, ManagerActivityLollipop.class);
                Bundle bundle = new Bundle();
                bundle.putSerializable("drawerItemQR", ManagerActivityLollipop.DrawerItem.SETTINGS);
                intent.putExtras(bundle);
                intent.putExtra("fromQR", true);
                intent.addFlags(Intent.FLAG_ACTIVITY_CLEAR_TOP);
                startActivity(intent);
                this.finish();
                break;
            }
            case R.id.qr_code_reset: {
                resetQR();
                break;
            }
            case R.id.qr_code_delete: {
                deleteQR();
                break;
            }
        }
        return super.onOptionsItemSelected(item);
    }

    @Override
    protected void onActivityResult(int requestCode, int resultCode, Intent intent) {
        super.onActivityResult(requestCode, resultCode, intent);
        if (requestCode == REQUEST_DOWNLOAD_FOLDER && resultCode == Activity.RESULT_OK && intent != null) {
            String parentPath = intent.getStringExtra(FileStorageActivityLollipop.EXTRA_PATH);
            if (parentPath != null) {
                File qrFile = null;
                if (megaApi == null) {
                    megaApi = ((MegaApplication) getApplication()).getMegaApi();
                }
                String myEmail = megaApi.getMyEmail();
                qrFile = buildQrFile(this,myEmail + "QRcode.jpg");
                if (qrFile == null) {
                    showSnackbar(drawerLayout, getString(R.string.general_error));
                }
                else {
                    if (qrFile.exists()) {
                        if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.M) {
                            boolean hasStoragePermission = (ContextCompat.checkSelfPermission(this, Manifest.permission.WRITE_EXTERNAL_STORAGE) == PackageManager.PERMISSION_GRANTED);
                            if (!hasStoragePermission) {
                                ActivityCompat.requestPermissions(this, new String[]{Manifest.permission.WRITE_EXTERNAL_STORAGE}, Constants.REQUEST_WRITE_STORAGE);
                            }
                        }

                        double availableFreeSpace = Double.MAX_VALUE;
                        try {
                            StatFs stat = new StatFs(parentPath);
                            availableFreeSpace = (double) stat.getAvailableBlocks() * (double) stat.getBlockSize();
                        } catch (Exception ex) {
                        }

                        if (availableFreeSpace < qrFile.length()) {
                            showSnackbar(drawerLayout, getString(R.string.error_not_enough_free_space));
                            return;
                        }
                        File newQrFile = new File(parentPath, myEmail + "QRcode.jpg");
                        if (newQrFile == null) {
                            showSnackbar(drawerLayout, getString(R.string.general_error));
                        }
                        else {
                            try {
                                newQrFile.createNewFile();
                                FileChannel src = new FileInputStream(qrFile).getChannel();
                                FileChannel dst = new FileOutputStream(newQrFile, false).getChannel();
                                dst.transferFrom(src, 0, src.size());       // copy the first file to second.....
                                src.close();
                                dst.close();
                                showSnackbar(drawerLayout, getString(R.string.success_download_qr, parentPath));
                            } catch (IOException e) {
                                showSnackbar(drawerLayout, getString(R.string.general_error));
                                e.printStackTrace();
                            }
                        }
                    }
                    else {
                        showSnackbar(drawerLayout, getString(R.string.error_download_qr));
                    }
                }
            }
        }
    }

    public void shareQR () {
        log("shareQR");

        if (myCodeFragment == null) {
            log("MyCodeFragment is NULL");
            myCodeFragment = (MyCodeFragment) qrCodePageAdapter.instantiateItem(viewPagerQRCode, 0);
        }
        if (myCodeFragment!= null && myCodeFragment.isAdded()){
            File qrCodeFile = myCodeFragment.queryIfQRExists();
            if (qrCodeFile != null && qrCodeFile.exists()){
                Intent share = new Intent(android.content.Intent.ACTION_SEND);
                share.setType("image/*");

                if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.N) {
                    log("Use provider to share");
                    Uri uri = FileProvider.getUriForFile(this, "mega.privacy.android.app.providers.fileprovider",qrCodeFile);
                    share.putExtra(Intent.EXTRA_STREAM, Uri.parse(uri.toString()));
                    share.addFlags(Intent.FLAG_GRANT_READ_URI_PERMISSION);
                }
                else{
                    share.putExtra(Intent.EXTRA_STREAM, Uri.parse("file://" + qrCodeFile));
                }
                startActivity(Intent.createChooser(share, getString(R.string.context_share)));
            }
            else {
                showSnackbar(drawerLayout, getString(R.string.error_share_qr));
            }
        }
    }

    public void resetQR () {
        log("resetQR");

        if (myCodeFragment == null) {
            log("MyCodeFragment is NULL");
            myCodeFragment = (MyCodeFragment) qrCodePageAdapter.instantiateItem(viewPagerQRCode, 0);
        }
        if (myCodeFragment != null && myCodeFragment.isAdded()) {
            myCodeFragment.resetQRCode();
        }
    }

    public void deleteQR () {
        log("deleteQR");

        if (myCodeFragment == null) {
            log("MyCodeFragment is NULL");
            myCodeFragment = (MyCodeFragment) qrCodePageAdapter.instantiateItem(viewPagerQRCode, 0);
        }
        if (myCodeFragment != null && myCodeFragment.isAdded()){
            myCodeFragment.deleteQRCode();
        }
    }

    public void resetSuccessfully (boolean success) {
        log("resetSuccessfully");
        if (success){
            showSnackbar(drawerLayout, getString(R.string.qrcode_reset_successfully));
        }
        else {
            showSnackbar(drawerLayout, getString(R.string.qrcode_reset_not_successfully));
        }
    }

    public void showSnackbar(View view, String s){
        if (view == null) {
            showSnackbar(Constants.SNACKBAR_TYPE, drawerLayout, s);
        }
        else {
            showSnackbar(Constants.SNACKBAR_TYPE, view, s);
        }
    }

    public static void log(String message) {
        Util.log("QRCodeActivity", message);
    }

    @Override
    public void onRequestStart(MegaApiJava api, MegaRequest request) {

    }

    @Override
    public void onRequestUpdate(MegaApiJava api, MegaRequest request) {

    }

    @Override
    public void onRequestFinish(MegaApiJava api, MegaRequest request, MegaError e) {
        log("onRequestFinish ");
        if(request.getType() == MegaRequest.TYPE_INVITE_CONTACT  && request.getNumber() == MegaContactRequest.INVITE_ACTION_ADD){
            if (scanCodeFragment == null) {
                log("ScanCodeFragment is NULL");
                scanCodeFragment = (ScanCodeFragment) qrCodePageAdapter.instantiateItem(viewPagerQRCode, 1);
            }
            if (scanCodeFragment != null && scanCodeFragment.isAdded()){
                scanCodeFragment.myEmail = request.getEmail();
                if (e.getErrorCode() == MegaError.API_OK){
                    log("OK INVITE CONTACT: "+request.getEmail());
                    scanCodeFragment.dialogTitleContent = R.string.invite_sent;
                    scanCodeFragment.dialogTextContent = R.string.invite_sent_text;
                    scanCodeFragment.showAlertDialog(scanCodeFragment.dialogTitleContent, scanCodeFragment.dialogTextContent, true);
                }
                else if (e.getErrorCode() == MegaError.API_EACCESS){
                    scanCodeFragment.dialogTitleContent = R.string.invite_not_sent;
                    scanCodeFragment.dialogTextContent = R.string.invite_not_sent_text_error;
                    scanCodeFragment.showAlertDialog(scanCodeFragment.dialogTitleContent, scanCodeFragment.dialogTextContent, false);
                }
                else if (e.getErrorCode() == MegaError.API_EEXIST){
                    scanCodeFragment.dialogTitleContent = R.string.invite_not_sent;
                    scanCodeFragment.dialogTextContent = R.string.invite_not_sent_text_already_contact;
                    scanCodeFragment.showAlertDialog(scanCodeFragment.dialogTitleContent, scanCodeFragment.dialogTextContent, true);
                }
                else if (e.getErrorCode() == MegaError.API_EARGS){
                    scanCodeFragment.dialogTitleContent = R.string.invite_not_sent;
                    scanCodeFragment.dialogTextContent = R.string.error_own_email_as_contact;
                    scanCodeFragment.showAlertDialog(scanCodeFragment.dialogTitleContent, scanCodeFragment.dialogTextContent, true);
                }
            }
        }
        //        megaApi.contactLinkQuery(request.getNodeHandle(), this);
        else if (request.getType() == MegaRequest.TYPE_CONTACT_LINK_QUERY){
            if (inviteContacts) {
                Intent intent = new Intent();
                intent.putExtra("mail", request.getEmail());
                setResult(RESULT_OK, intent);
                finish();
            }
            else {
                if (scanCodeFragment == null) {
                    log("ScanCodeFragment is NULL");
                    scanCodeFragment = (ScanCodeFragment) qrCodePageAdapter.instantiateItem(viewPagerQRCode, 1);
                }
                if (scanCodeFragment != null && scanCodeFragment.isAdded()) {
                    scanCodeFragment.myEmail = request.getEmail();
                    scanCodeFragment.initDialogInvite(request, e);
                }
            }
        }
        else if (request.getType() == MegaRequest.TYPE_GET_ATTR_USER) {
            if (scanCodeFragment == null) {
                log("ScanCodeFragment is NULL");
                scanCodeFragment = (ScanCodeFragment) qrCodePageAdapter.instantiateItem(viewPagerQRCode, 1);
            }
            if (scanCodeFragment != null && scanCodeFragment.isAdded()) {
                if (e.getErrorCode() == MegaError.API_OK) {
                    log("Get user avatar OK");
                    scanCodeFragment.setAvatar();
                } else {
                    log("Get user avatar FAIL");
                    scanCodeFragment.setDefaultAvatar();
                }
            }
        }
//        megaApi.contactLinkCreate(true/false, this);
        else if (request.getType() == MegaRequest.TYPE_CONTACT_LINK_CREATE) {
            if (myCodeFragment == null){
                log("MyCodeFragment is NULL");
                myCodeFragment = (MyCodeFragment) qrCodePageAdapter.instantiateItem(viewPagerQRCode, 0);
            }
            if (myCodeFragment != null && myCodeFragment.isAdded()){
                myCodeFragment.initCreateQR(request, e);
            }
        }
//        megaApi.contactLinkDelete(request.getNodeHandle(), this);
        else if (request.getType() == MegaRequest.TYPE_CONTACT_LINK_DELETE){
            if (myCodeFragment == null){
                log("MyCodeFragment is NULL");
                myCodeFragment = (MyCodeFragment) qrCodePageAdapter.instantiateItem(viewPagerQRCode, 0);
            }
            if (myCodeFragment != null && myCodeFragment.isAdded()){
                myCodeFragment.initDeleteQR(request, e);
            }
        }
    }

    @Override
    public void onRequestTemporaryError(MegaApiJava api, MegaRequest request, MegaError e) {

    }

    public void deleteSuccessfully() {
        log("deleteSuccessfully");
        shareMenuItem.setVisible(false);
        saveMenuItem.setVisible(false);
        settingsMenuItem.setVisible(true);
        resetQRMenuItem.setVisible(true);
        deleteQRMenuItem.setVisible(false);
    }

    public void createSuccessfully() {
        log("createSuccesfully");
        shareMenuItem.setVisible(true);
        saveMenuItem.setVisible(true);
        settingsMenuItem.setVisible(true);
        resetQRMenuItem.setVisible(true);
        deleteQRMenuItem.setVisible(true);
    }
}

File: app/src/main/java/mega/privacy/android/app/presentation/qrcode/QRCodeActivity.kt
package mega.privacy.android.app.presentation.qrcode

import android.Manifest
import android.content.Intent
import android.media.MediaScannerConnection
import android.net.Uri
import android.os.Bundle
import android.os.StatFs
import android.view.View
import androidx.activity.compose.setContent
import androidx.activity.viewModels
import androidx.compose.runtime.LaunchedEffect
import androidx.compose.runtime.collectAsState
import androidx.compose.runtime.getValue
import androidx.core.content.ContextCompat
import androidx.core.content.FileProvider
import androidx.lifecycle.Lifecycle
import androidx.lifecycle.compose.collectAsStateWithLifecycle
import com.google.accompanist.permissions.ExperimentalPermissionsApi
import com.google.accompanist.permissions.isGranted
import com.google.accompanist.permissions.rememberPermissionState
import dagger.hilt.android.AndroidEntryPoint
import mega.privacy.android.app.R
import mega.privacy.android.app.activities.PasscodeActivity
import mega.privacy.android.app.arch.extensions.collectFlow
import mega.privacy.android.app.main.FileStorageActivity
import mega.privacy.android.app.modalbottomsheet.ModalBottomSheetUtil.isBottomSheetDialogShown
import mega.privacy.android.app.modalbottomsheet.QRCodeSaveBottomSheetDialogFragment
import mega.privacy.android.app.presentation.extensions.isDarkMode
import mega.privacy.android.app.presentation.qrcode.model.MyQRTab
import mega.privacy.android.app.presentation.qrcode.mycode.MyCodeFragment
import mega.privacy.android.app.presentation.qrcode.mycode.MyCodeViewModel
import mega.privacy.android.app.presentation.qrcode.scan.ScanCodeViewModel
import mega.privacy.android.app.presentation.settings.SettingsActivity
import mega.privacy.android.app.presentation.settings.model.TargetPreference
import mega.privacy.android.app.utils.CacheFolderManager
import mega.privacy.android.app.utils.Constants
import mega.privacy.android.app.utils.Util
import mega.privacy.android.app.utils.permission.PermissionUtils
import mega.privacy.android.shared.theme.MegaAppTheme
import mega.privacy.android.domain.entity.ThemeMode
import mega.privacy.android.domain.usecase.GetThemeMode
import timber.log.Timber
import java.io.File
import java.io.FileInputStream
import java.io.FileOutputStream
import java.io.IOException
import javax.inject.Inject

/**
 * QR code Activity
 */
@AndroidEntryPoint
class QRCodeActivity : PasscodeActivity() {

    /**
     * Get theme mode
     */
    @Inject
    lateinit var getThemeMode: GetThemeMode

    private val qrCodeViewModel: MyCodeViewModel by viewModels()
    private val scanCodeViewModel: ScanCodeViewModel by viewModels()

    private lateinit var rootView: View

    private var inviteContacts = false
    private var showScanQrView = false

    private var qrCodeSaveBottomSheetDialogFragment: QRCodeSaveBottomSheetDialogFragment? = null

    @OptIn(ExperimentalPermissionsApi::class)
    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)

        savedInstanceState?.let {
            showScanQrView = it.getBoolean(Constants.OPEN_SCAN_QR, false)
            inviteContacts = it.getBoolean(Constants.INVITE_CONTACT, false)
        } ?: run {
            showScanQrView = intent.getBooleanExtra(Constants.OPEN_SCAN_QR, false)
            inviteContacts = intent.getBooleanExtra(Constants.INVITE_CONTACT, false)
        }

        setContent {
            val mode by getThemeMode()
                .collectAsStateWithLifecycle(initialValue = ThemeMode.System)
            val qrCodeUIState by qrCodeViewModel.uiState.collectAsState()
            MegaAppTheme(isDark = mode.isDarkMode()) {

                val cameraPermissionState = rememberPermissionState(
                    Manifest.permission.CAMERA
                )

                if (!cameraPermissionState.status.isGranted) {
                    LaunchedEffect(Unit) {
                        cameraPermissionState.launchPermissionRequest()
                    }
                }
                QRCodeBody(
                    qrCodeUIState = qrCodeUIState,
                    onBackPressed = { onBackPressedDispatcher.onBackPressed() },
                    onDeleteQRCode = ::deleteQR,
                    onResetQRCode = ::resetQR,
                    onGotoSettings = ::gotoSettings,
                    onSaveQRCode = ::saveQR,
                    onShareQRCode = { qrCodeViewModel.startSharing() },
                    initialTab = if (showScanQrView || inviteContacts) MyQRTab.ScanQRCode else MyQRTab.MyQRCode
                )
            }
        }
        setupFlow()
        rootView = findViewById(android.R.id.content)
        scanCodeViewModel.updateFinishActivityOnScanComplete(inviteContacts)
    }

    @Deprecated("Deprecated in Java")
    @Suppress("deprecation")
    override fun onActivityResult(requestCode: Int, resultCode: Int, intent: Intent?) {
        super.onActivityResult(requestCode, resultCode, intent)

        intent?.takeIf { requestCode == Constants.REQUEST_DOWNLOAD_FOLDER && resultCode == RESULT_OK }
            ?.getStringExtra(FileStorageActivity.EXTRA_PATH)
            ?.let { parentPath -> saveToFileSystem(parentPath) }
    }

    override fun onSaveInstanceState(outState: Bundle) {
        super.onSaveInstanceState(outState)
        outState.putBoolean(Constants.OPEN_SCAN_QR, showScanQrView)
        outState.putBoolean(Constants.INVITE_CONTACT, inviteContacts)
    }

    private fun setupFlow() {
        collectFlow(qrCodeViewModel.uiState, Lifecycle.State.RESUMED) { uiState ->
            with(uiState) {
                localQRCodeFile?.let {
                    shareQR()
                    qrCodeViewModel.finishSharing()
                }
            }
        }
    }

    private fun saveToFileSystem(parentPath: String) {
        val myEmail = megaApi.myEmail
        val qrFile: File? = CacheFolderManager.buildQrFile(
            fileName = myEmail + MyCodeFragment.QR_IMAGE_FILE_NAME
        )

        if (qrFile == null) {
            showSnackbar(rootView, getString(R.string.general_error))
            return
        }

        if (!qrFile.exists()) {
            showSnackbar(rootView, getString(R.string.error_download_qr))
            return
        }

        val hasStoragePermission =
            PermissionUtils.hasPermissions(this, Manifest.permission.WRITE_EXTERNAL_STORAGE)
        if (!hasStoragePermission) {
            PermissionUtils.requestPermission(
                this,
                Constants.REQUEST_WRITE_STORAGE,
                Manifest.permission.WRITE_EXTERNAL_STORAGE
            )
        }

        val availableFreeSpace = try {
            StatFs(parentPath).run {
                availableBlocksLong.toDouble() * blockSizeLong.toDouble()
            }
        } catch (e: Exception) {
            Timber.e(e)
            Double.MAX_VALUE
        }

        if (availableFreeSpace < qrFile.length()) {
            showSnackbar(
                rootView,
                getString(R.string.error_not_enough_free_space)
            )
            return
        }

        val newQrFile =
            File(parentPath, "$myEmail${MyCodeFragment.QR_IMAGE_FILE_NAME}")

        // For Android 11+ device, force to refresh MediaStore. Otherwise it is possible
        // that target file cannot be written.
        if (Util.isAndroid11OrUpper()) {
            MediaScannerConnection.scanFile(
                this,
                arrayOf(newQrFile.absolutePath),
                arrayOf("image/jpeg"),
                null
            )
        }

        try {
            newQrFile.createNewFile()
            val src = FileInputStream(qrFile).channel
            val dst = FileOutputStream(newQrFile, false).channel
            dst.transferFrom(src, 0, src.size())
            src.close()
            dst.close()
            showSnackbar(
                rootView,
                getString(R.string.success_download_qr, parentPath)
            )
        } catch (e: IOException) {
            Timber.e(e)
            showSnackbar(
                rootView,
                getString(R.string.general_error)
            )
        }
        return
    }

    private fun resetQR() {
        Timber.d("resetQRCode")
        val penColor = ContextCompat.getColor(this, R.color.dark_grey)
        val bgColor = ContextCompat.getColor(this, R.color.white_grey_700)
        val avatarBorderColor = ContextCompat.getColor(this, R.color.white_dark_grey)
        qrCodeViewModel.resetQRCode(
            width = MyCodeFragment.QRCODE_WIDTH,
            height = MyCodeFragment.QRCODE_WIDTH,
            penColor = penColor,
            bgColor = bgColor,
            avatarWidth = MyCodeFragment.AVATAR_WIDTH,
            avatarBorderWidth = MyCodeFragment.AVATAR_BORDER_WIDTH,
            avatarBorderColor = avatarBorderColor,
        )
    }

    private fun deleteQR() {
        Timber.d("deleteQR")
        qrCodeViewModel.deleteQR()
    }

    private fun gotoSettings() {
        val settingsIntent = SettingsActivity.getIntent(this, TargetPreference.QR).apply {
            addFlags(Intent.FLAG_ACTIVITY_CLEAR_TOP)
        }
        startActivity(settingsIntent)
        finish()
    }

    private fun saveQR() {
        if (!qrCodeSaveBottomSheetDialogFragment.isBottomSheetDialogShown()) {
            qrCodeSaveBottomSheetDialogFragment =
                QRCodeSaveBottomSheetDialogFragment().also {
                    it.show(supportFragmentManager, it.tag)
                }
        }
    }

    private fun shareQR() {
        Timber.d("shareQR")

        qrCodeViewModel.uiState.value.localQRCodeFile?.let { qrCodeFile ->
            Timber.d("Use provider to share")

            val uri = FileProvider.getUriForFile(
                this@QRCodeActivity,
                Constants.AUTHORITY_STRING_FILE_PROVIDER,
                qrCodeFile
            )
            val shareIntent = Intent(Intent.ACTION_SEND).apply {
                type = "image/*"
                putExtra(Intent.EXTRA_STREAM, Uri.parse(uri.toString()))
                addFlags(Intent.FLAG_GRANT_READ_URI_PERMISSION)
            }

            startActivity(Intent.createChooser(shareIntent, getString(R.string.context_share)))
        } ?: run {
            showSnackbar(rootView, getString(R.string.error_share_qr))
        }

    }
}
